### PR TITLE
gh-112087: Update list impl to be thread-safe with manual CS

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-112087

### CODE REVIEW

**Comment by corona10** [comment]
The rest of the parts can be separated into two parts.
A. xxx_lock implementations.
B. Where PyList_FetchItem is needed.

A is a headache; atomic operation with default build will require some overhead than pure operation; not sure how we can solve this with a seamless approach.

B is quite simple; if we already have similar APIs, then we can just use them. If not, we should introduce one.

File: Objects/listobject.c [hunk 1]
```diff
@@ -2632,7 +2652,12 @@ PyList_AsTuple(PyObject *v)
         PyErr_BadInternalCall();
         return NULL;
     }
-    return _PyTuple_FromArray(((PyListObject *)v)->ob_item, Py_SIZE(v));
+    PyObject *ret;
+    PyListObject *self = (PyListObject *)v;
```
Comment (colesbury) [inline Objects/listobject.c:2658]: GHA is complaining about `self` being unused.

File: Objects/listobject.c [hunk 2]
```diff
@@ -267,15 +267,22 @@ PyList_SetItem(PyObject *op, Py_ssize_t i,
...
         Py_XDECREF(newitem);
         PyErr_SetString(PyExc_IndexError,
                         "list assignment index out of range");
-        return -1;
+        ret = -1;
+        goto end;
     }
-    p = ((PyListObject *)op) -> ob_item + i;
+    p = self->ob_item + i;
     Py_XSETREF(*p, newitem);
```
Comment (colesbury) [inline Objects/listobject.c:281]: We will want to change the `Py_XSETREF()` call in the future when we add the thread-safe implementation for `PyList_FetchItem()` (`PyList_GetItemRef()`?)

**Review by colesbury — APPROVED** [review]
Looks good other than the unused variable.

I scanned through listobject.c and it looks like there are still other functions that should be fairly straightforward to make thread-safe, such as list_copy_impl and list_append.

What do you mean by "xxx_lock implementations"?

File: Objects/listobject.c (hunk 2 above)
Comment (corona10) [reply Objects/listobject.c:281]: Yeah it should be

**Comment by corona10** [comment]
@colesbury Updated PTAL :)

**Comment by corona10** [comment]
> What do you mean by "xxx_lock implementations"?

It was about list_item_locked things that you updated implementations to use _Py_atomic_store/load operations.

**Review by colesbury — APPROVED** [review]
LGTM.

I wonder if it's worth doing something to abstract out the locking in type slots (like `tp_richcompare`).

**Comment by corona10** [comment]
> I wonder if it's worth doing something to abstract out the locking in type slots (like tp_richcompare).

It might be up to case, `list_richcompare` was too complex case to handle without extracting core logics.

**Pull request merged**